### PR TITLE
RavenDB-21104 - Failover of resharding not working properly

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21104/Failover-of-resharding-not-working-properly

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentsTransaction.cs [hunk 1]
```diff
@@ -172,5 +177,10 @@ internal void CheckIfShouldDeleteAttachmentStream(Slice hash)
             _attachmentHashesToMaybeDelete ??= new();
             _attachmentHashesToMaybeDelete.Add(clone);
         }
+
+        internal void InvokeDocumentsMigration(Task migrationTask)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentsTransaction.cs]:
```csharp
void ExecuteDocumentsMigrationBeforeCommit()
{
    _executeDocumentsMigrationBeforeCommit = true;
}
````

Comment (ppekrol) [reply src/Raven.Server/Documents/DocumentsTransaction.cs]: and then if this is true, execute `_context.DocumentsDatabase.DocumentsMigrator.ExecuteMoveDocumentsAsync();`

Comment (shiranshalom) [reply src/Raven.Server/Documents/DocumentsTransaction.cs]: done

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs [hunk 2]
```diff
@@ -15,8 +16,11 @@ public ShardedDocumentPutAction(ShardedDocumentsStorage documentsStorage, Sharde
     }
 
     // TODO need to make sure we check that for counters/TS/etc...
-    public override void ValidateId(Slice lowerId)
+    public override void ValidateId(DocumentsOperationContext context, Slice lowerId, DocumentFlags documentFlags = DocumentFlags.None)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs:19]: what about deletion of a document?

File: src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs [hunk 3]
```diff
@@ -26,6 +30,15 @@ public override void ValidateId(Slice lowerId)
             if (_documentDatabase.ForTestingPurposes != null && _documentDatabase.ForTestingPurposes.EnableWritesToTheWrongShard)
                 return;
 
+            if (documentFlags.Contain(DocumentFlags.FromReplication))
+            {
+                // RavenDB-21104
+                // we allow writing the document to the wrong shard to avoid inconsistent data within the shard group
+                // and handle the leftovers at the end of the transaction 
+                context.Transaction.ExecuteDocumentsMigrationBeforeCommit();
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs:38]: shouldn't this be _after_ commit? 
what will happened if the task was completed before the actual commit?

File: src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs:19]: done in https://github.com/ravendb/ravendb/pull/17060

File: src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentPutAction.cs:38]: done in https://github.com/ravendb/ravendb/pull/17202

**Pull request merged**